Fix handling of unsigned types in numeric validators

vadz opened the pull request:
This is still a bit ugly and we probably should get rid of all this `LongestValueType` stuff completely and just work with strings as "value containers" instead, but it's not more complicated than before (I'd even say it's slightly simpler) and does allow the new unit tests to pass, so seems like an improvement.

@AliKet please let me know if you see anything wrong with this.

vadz committed:
Get rid of CppUnit boilerplate in numeric validator unit tests

No real changes, just drop CppUnit::TestCase inheritance and the legacy
macros and use TEST_CASE_METHOD() instead.

vadz committed:
Replace CPPUNIT_ASSERT() macros in wxNumberFormatter unit tests

Just use CHECK() rather than CPPUNIT_ASSERT_EQUAL() etc.

vadz committed:
Replace CPPUNIT_ASSERT() macros in numeric validator unit tests

Simplify code and use CHECK() rather than REQUIRE(), which is what
CPPUNIT_ASSERT() is defined as, to continue running the tests even if
some of them fail.

vadz committed:
Don't define copy ctor in numeric validator classes unnecessarily

There is just no need to do it manually, when the compiler can do the
job perfectly well itself.

No real changes, but this will simplify the upcoming refactoring.

vadz committed:
Fix confusion around '-' handling in numeric validators code

Actual implementation of IsCharOk() didn't correspond to the comments in
it or near the function declaration which stated that it's never called
with ch='-' as argument -- it was called with it and called IsMinusOk()
right before a comment saying that it doesn't need to do it, which was
very confusing.

Fix this by making the behaviour really correspond to the comments and
handling '-' at the base class level. This required introducing a new
pure virtual CanBeNegative() function, but it's going to be useful for
other things later too.

Still keep IsMinusOk() helper, but make it private now because it
doesn't need to be called from the derived class IsCharOk() any longer.

vadz committed:
Document wxNumberFormatter overloads using long long

These overloads were added back in f2a5052baa (Add support for long long
to wxNumberFormatter., 2011-01-19) but never documented.

vadz committed:
Get rid of CppUnit boilerplate in wxNumberFormatter unit tests

No real changes, just drop CppUnit::TestCase inheritance and the legacy
macros and use TEST_CASE_METHOD() instead.

vadz committed:
Make wxNumberFormatter tests more concise and readable

Define a couple of helper functions to avoid over long lines or having
to break them and also align the tests vertically to allow scanning them
more easily.

No real changes.

vadz committed:
Add support for unsigned long long to wxNumberFormatter

This is necessary in order to deal with the numbers greater than
wxINT64_MAX that can't be represented in just long long.

It also allows to implement the intuitive handling of minus sign for the
unsigned numbers, i.e. not to accept it in FromString(), unlike the
standard functions which do (and parse -1 as 0xffff...fff).

Also extend the tests to check for more boundary cases.

vadz committed:
Fix parsing unsigned numbers in wxIntegerValidatorBase

Use wxNumberFormatter::FromString() overload taking wxULongLong_t to
allow parsing numbers greater than LLONG_MAX.

vadz committed:
Use min/max values of correct type in numeric validators

Use the actual type of the value, not LongestValueType, for storing
m_min and m_max as this is necessary for the comparisons between the
value and them to work correctly for unsigned types.

Also check for precision loss when converting from the bigger
LongestValueType to the actual type.

Add new unit tests, that failed before, but pass now.

AliKet commented on src/common/valnum.cpp:
```diff
@@ -234,9 +234,29 @@ wxString wxIntegerValidatorBase::ToString(LongestValueType value) const
+        wxULongLong_t uvalue;
+#else
+        wxULongLong_t uvalue;
+#endif
```
Looks like we should do this instead:
```
#ifdef wxHAS_LONG_LONG_T_DIFFERENT_FROM_LONG
        wxULongLong_t uvalue;
#else
        unsigned long uvalue;
#endif
```
to be consistent with what is already done by the library itself elsewhere!

vadz replied:
Oops, yes, this is what this was supposed to be, thanks for noticing.

AliKet commented on src/common/valnum.cpp:
```diff
@@ -234,9 +234,29 @@ wxString wxIntegerValidatorBase::ToString(LongestValueType value) const
 }
```
`wxIntegerValidatorBase::ToString` should be updated too:

```
{
    if ( CanBeNegative() )
        return wxNumberFormatter::ToString(value, GetFormatFlags());

    wxULongLong_t uvalue = static_cast<wxULongLong_t>(value);
    return wxNumberFormatter::ToString(uvalue, GetFormatFlags());
}
```
Otherwise `wxNumValidatorBase::OnKillFocus` would leave the control with a negative value after normalizing its content.

vadz replied:
Hmm, how exactly is this going to happen? I'd like to add a unit test for this, but I'm not sure what exactly needs to be done to trigger this, do you know?

AliKet replied:
The problem is at this [line](https://github.com/wxWidgets/wxWidgets/blob/master/include/wx/valnum.h#L257),
i.e. `s = this->ToString(value);`

`OnKillFocus` calls `NormalizeString` which calls `NormalizeValue` thinking they are dealing with a signed value! (i.e. `LongestValueType`).

Reproduced by this:
```
TEST_CASE_METHOD(NumValidatorTestCase, "ValNum::TransferULL", "[valnum]")
{
...
    CHECK( valULL.TransferToWindow() );
    CHECK( m_text->GetValue() == "18446744073709551615" ); // <-- fails!
}
```

vadz replied:
Yes, you're absolutely right, sorry for not realizing it sooner. Fixed now, with more tests to check for it, thanks again!

AliKet commented on include/wx/numformatter.h:
```diff
@@ -34,6 +34,8 @@ class WXDLLIMPEXP_BASE wxNumberFormatter
                              int style = Style_WithThousandsSep);
 #endif // wxHAS_LONG_LONG_T_DIFFERENT_FROM_LONG
+    static wxString ToString(wxULongLong_t val,
+                             int style = Style_WithThousandsSep);
```
Shouldn't this be inside the wxHAS_LONG_LONG_T_DIFFERENT_FROM_LONG prep. ?

vadz replied:
No because `unsigned [long] long` is always a different type, even if `long == long long`.

AliKet commented on tests/strings/numformatter.cpp:
```diff
@@ -224,10 +228,63 @@ TEST_CASE_METHOD(NumFormatterTestCase, "NumFormatter::LongLongFromString", "[num
+    CHECK( u == 1234567 );
+
+    CHECK( wxNumberFormatter::FromString("9223372036854775807", &u) );
+    CHECK( u == wxINT64_MAX );
```
`CHECK( u == static_cast<wxULongLong_t>(wxINT64_MAX) );`

vadz replied:
Thanks for noticing this, fixed now.

AliKet commented on tests/validators/valnum.cpp:
```diff
@@ -96,13 +96,42 @@ TEST_CASE_METHOD(NumValidatorTestCase, "ValNum::TransferUnsigned", "[valnum]")
+
+    m_text->ChangeValue("9223372036854775807"); // == LLONG_MAX
+    CHECK( valULL.TransferFromWindow() );
+    CHECK( value == wxINT64_MAX );
```
`    CHECK( value == static_cast<wxULongLong_t>(wxINT64_MAX) );`

AliKet commented:
Thank you for taking care of this and for the cleanup!

Looks good to me. thanks.

AliKet commented:
BTW, I cannot reproduce this failure (for some) locally:
```
strings/numformatter.cpp:209: FAILED:
  CHECK( !wxNumberFormatter::FromString("-", &l) )
with expansion:
  false
```
... and I have no idea why this is happening!

vadz committed:
Use CHECK_FALSE in wxNumberFormatter unit tests

Just replace CHECK(!condition) with CHECK_FALSE(condition), this should
result in slightly more clear error messages.

Also make tests for long and long long more consistent.

vadz commented:
Thanks for looking at this!

There is a weird [test failure](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/37879480/job/t7lcqqcjna17iqru#L1343) with parsing "-" using MinGW 5.3 but I'm probably going to just disable this test for it, this looks like an actual problem with its CRT implementation and doesn't happen with later versions.

vadz committed:
Relax check for converting "-" to long long in the unit tests

This somehow succeeds when using gcc 4.8 under Ubuntu 14.04 or MinGW
5.3, so don't fail the tests in this case, but still warn about it
because it seems quite unexpected.

vadz committed:
Add wxIntegerValidatorBase::ULongestValueType typedef

Allow just writing ULongestValueType instead of writing #ifdefs in
several places.

vadz committed:
Fix transferring unsigned values to the window too

In addition to not accepting "-1" as a valid unsigned value, we also
must not format unsigned values as signed ones, i.e. we need to use
wxNumberFormatter::ToString() overload taking unsigned for them.

Do it in wxIntegerValidatorBase::ToString(), used by TransferToWindow(),
and add more tests for the latter.

Also reorganize the tests in sections and use REQUIRE() for the checks
that should prevent the rest of the section from running if they fail.

AliKet commented on tests/validators/valnum.cpp:
```diff
@@ -22,55 +22,29 @@
-void NumValidatorTestCase::tearDown()
+NumValidatorTestCase::~NumValidatorTestCase()
 {
-    wxTheApp->GetTopWindow()->DestroyChildren();
```
Sorry, for forgetting about this, but
`NumValidatorTestCase :: Interactive()` creates a wxTextCtrl which should be explicitly deleted after this modification.

vadz replied:
Thanks for noticing this! I'll add `wxON_BLOCK_EXIT_OBJ` after `text2` creation.
